src/lib/util/chdcd.cpp

```diff
@@ -1129,6 +1134,437 @@ chd_error chdcd_parse_cue(const char *tocfname, cdrom_toc &outtoc, chdcd_track_i
+					outinfo.track[trknum].offset = 0;
+				}
+			}
+			else
+			{
+				/* if we have the same filename as the next track, do it that way */
+				if (outinfo.track[trknum].fname.compare(outinfo.track[trknum+1].fname)==0)
+				{
+					outtoc.tracks[trknum].frames = outinfo.track[trknum+1].idx0offs - outinfo.track[trknum].idx0offs;
+
+					if (trknum == 0)    // track 0 offset is 0
+					{
+						outinfo.track[trknum].offset = 0;
+					}
+					else
+					{
+						outinfo.track[trknum].offset = outinfo.track[trknum-1].offset + outtoc.tracks[trknum-1].frames * (outtoc.tracks[trknum-1].datasize + outtoc.tracks[trknum-1].subsize);
+					}
+
+					if (!outtoc.tracks[trknum].frames)
+					{
+						printf("ERROR: unable to determine size of track %d, missing INDEX 01 markers?\n", trknum+1);
+						return CHDERR_INVALID_DATA;
+					}
+				}
+				else    /* data files are different */
+				{
+					tlen = get_file_size(outinfo.track[trknum].fname.c_str());
+					if (tlen == 0)
+					{
+						printf("ERROR: couldn't find bin file [%s]\n", outinfo.track[trknum].fname.c_str());
+						return CHDERR_FILE_NOT_FOUND;
+					}
+					tlen /= (outtoc.tracks[trknum].datasize + outtoc.tracks[trknum].subsize);
+					outtoc.tracks[trknum].frames = tlen;
+					outinfo.track[trknum].offset = 0;
+				}
+			}
+		}
+
+		/*
+		 * This is the special handling for Dreamcast GDI
+		 *
+		 * - high density image always starts at physical offset 45000
+		 * - last track of single density image is padded out
+		 * - physical offset is stored for every track
+		 * - pregap and pgtype is zeroed
+		 *
+		 * I think this last step is wrong but it's what chdcd_parse_gdi does and flycast expects it
+		 */
+		if (trknum != 0)
+		{
+			if (outtoc.tracks[trknum].multicuearea == HIGH_DENSITY && outtoc.tracks[trknum-1].multicuearea == SINGLE_DENSITY) {
+				outtoc.tracks[trknum].physframeofs = 45000;
+			} else {
+				outtoc.tracks[trknum].physframeofs = outtoc.tracks[trknum-1].physframeofs + outtoc.tracks[trknum-1].frames;
+			}
+
+			int dif=outtoc.tracks[trknum].physframeofs-(outtoc.tracks[trknum-1].frames+outtoc.tracks[trknum-1].physframeofs);
+			outtoc.tracks[trknum-1].frames += dif;
+			outtoc.tracks[trknum-1].padframes = dif;
+
+		}
+
+		// flycast only works when these are zeroed
+		outtoc.tracks[trknum].pregap = 0;
+		outtoc.tracks[trknum].pgtype = 0;
+	}
+
+#if 0
+	for (trknum = 0; trknum < outtoc.numtrks; trknum++)
+	{
+		printf("trk %d: %d frames @ offset %d, pad=%d, area=%d, phys=%d, pre=%d, post=%d, idx0=%d, idx1=%d, (true %d)\n",
+			trknum+1,
+			outtoc.tracks[trknum].frames,
+			outinfo.track[trknum].offset,
+			outtoc.tracks[trknum].padframes,
+			outtoc.tracks[trknum].multicuearea,
+			outtoc.tracks[trknum].physframeofs,
+			outtoc.tracks[trknum].pregap,
+			outtoc.tracks[trknum].postgap,
+			outinfo.track[trknum].idx0offs,
+			outinfo.track[trknum].idx1offs,
+			outtoc.tracks[trknum].frames - outtoc.tracks[trknum].padframes);
+	}
+#endif
+
+	/* close the input TOC */
```

Comment:
This comment seems to be misplaced.

Reply:
Resolved in 91af4f306760585ffc48a0bf302144cd81823cd8